functors_task_2/student/src/impl.cpp

```diff
@@ -7,48 +7,54 @@ void DataBrowser::userLeave(const std::string &userId)
 
 bool DataBrowser::getDataType1(const std::string &userId, std::vector<size_t> &returnValues) const
 {
-    const auto& it = m_dataReaders.find(userId);
-    if (it == m_dataReaders.cend())
+    auto fn = [&](const std::unique_ptr<IDataSelector>& selector) 
     {
-        return false;
-    }
-    if (it->second == nullptr)
-    {
-        return false;
-    }
-    return it->second->getDataType1(returnValues, 0);
+        return invokeDataRequest(
+            [](const IDataSelector* selector, std::vector<size_t> &returnValues) 
+            {
+                return selector->getDataType1(returnValues, 0);
+            }, 
+            selector, returnValues
+        );
+    };
+    
+    return safeCall(userId, fn);
 }
 
 bool DataBrowser::getDataType2(std::vector<size_t> &returnValues, const std::string &userId) const
 {
-    const auto& it = m_dataReaders.find(userId);
-    if (it == m_dataReaders.cend())
-    {
-        return false;
-    }
-    if (it->second == nullptr)
+    auto fn = [&](const std::unique_ptr<IDataSelector>& selector) 
     {
-        return false;
-    }
-    return it->second->getDataType2(returnValues);
+        auto lambda = [](std::vector<size_t> &returnValues, const IDataSelector* selector) 
+        {
+            return selector->getDataType2(returnValues);
+        };
+
+        auto boundFn = std::bind(lambda, std::placeholders::_2, std::placeholders::_1);
```

Comment:
Hm, looks interesting if it was intended, but not sure we need to complicate things, arguments order may be changed in lambda itself
But thanks for showing example with bind